README.md
lxadm 0.1.6
============
Manage illumos LX zones.

[![Build Status](https://travis-ci.org/hadfl/lxadm.svg?branch=master)](https://travis-ci.org/hadfl/lxadm)

`lxadm` takes care of setting up LX zones on illumos derived operating systems.

Setup
-----

`lxadm` comes as a prebuilt pure perl package, so it should install out of
the box on any machine with a current perl installation.

```sh
wget https://github.com/hadfl/lxadm/releases/download/v0.1.6/lxadm-0.1.6.tar.gz
tar zxvf lxadm-0.1.6.tar.gz
cd lxadm-0.1.6
./configure --prefix=/opt/lxadm-0.1.6 
```

Now you can run

```sh
gmake
gmake install
```

Check the [man page](doc/lxadm.pod) for information about how to use lxadm.

Support and Contributions
-------------------------
If you find a problem with `lxadm`, please open an Issue on GitHub.

And if you have a contribution, please send a pull request.

Enjoy!

Dominik Hassler & Tobi Oetiker
2017-12-11


CHANGES
commit 139c2408d24263afcf404c0d9913febe577516ad
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Dec 11 15:31:44 2017 +0100

    lx brand filter

commit 9a492c76282758f9f9c221b6c360943c3972ba4e
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sun Oct 1 17:06:44 2017 +0200

    ready for 0.1.5

commit 09a5bbdc1f6bc0445b83d5f940d46f767041bef9
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sun Oct 1 16:55:34 2017 +0200

    remove lx brand check

commit a324f641eeb18d906fd324ae4f5f7bf890dbf2ad
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sun Oct 1 16:45:03 2017 +0200

    build now allows setting of var data dir

commit 561cc65baf33f6ac155c21334f7cc34926d3f9d8
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Sep 29 15:34:17 2017 +0200

    fix readme

commit 4f0908401444ee51cd24cb8b13ef4e2fc8ee499a
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sat Aug 19 14:15:37 2017 +0200

    ready for 0.1.4

commit e83702a49a026140850bfbc99e84178f1821e9db
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sat Aug 19 14:04:31 2017 +0200

    make 'net' resource optional

commit af385e44bff2b04988c9bfa9c5c97d0a998742c2
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Tue Jan 24 10:05:17 2017 +0100

    ready for 0.1.3

commit 67141922732dc671e29502a9aadab40108f79269
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Tue Jan 24 09:52:03 2017 +0100

    - remove empty attr resources

commit b991ba5dc9b462bc9d345baed424e149a7e479dc
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Fri Dec 16 17:03:50 2016 +0100

    ready for 0.1.2

commit 30dacac13a58814f53b4050601e941a78cca57e0
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Fri Dec 16 16:32:52 2016 +0100

    - use Illumos::Zones 0.1.4

commit 928870755508610bba60bf11186929a175a5968a
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Fri Dec 16 16:07:17 2016 +0100

    - proper rctl handling by aliases

commit 5f2182ae720cdab9263f35d309e079e6c61da3f8
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Fri Dec 16 15:03:09 2016 +0100

    - check if zone is a lx zone

commit d803069818b0a5e9b523d3501453c70efde3872d
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Thu Dec 15 19:25:54 2016 +0100

    - typos

commit aee02ac54616bbf1a18c5359352dab3f841347cf
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Thu Dec 15 15:59:56 2016 +0100

    - git attributes, perl 5.22

commit 85fe41c4ac9cd3952a468ee27cb7917a593f8128
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Thu Dec 15 15:48:15 2016 +0100

    - gitattributes added

commit 94c16cffbef933c943feac3e28795e8a08bf8727
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Thu Dec 15 15:32:51 2016 +0100

    ready for 0.1.1

commit c214e09efa01b34f128c3c2106bda8ce3072c5a7
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Thu Dec 15 14:50:43 2016 +0100

    - fix dependencies

commit 82e1d6335b998e46c8d44852238e111e84fdca0c
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Wed Dec 14 16:31:22 2016 +0100

    - docu fixes

commit 690fc6b53693602e65142f73d7781cd14b5197d3
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Wed Dec 14 15:12:39 2016 +0100

    - gitter again

commit 3dfe2d424d259638e7f79f462604ffc38ad746d1
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Wed Dec 14 14:02:30 2016 +0100

    - remove gitter for now

commit eb02abd836e5781d9c8a1bd380e6ca81205e9b03
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Tue Dec 13 19:39:36 2016 +0100

    - add gitter

commit 69c6033c773c9ebe8a4bf39d2303f412f5312bb9
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Tue Dec 13 16:56:53 2016 +0100

    - fix travis

commit d31cbd333c90969cb4faa3c466735731ef92bb20
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Tue Dec 13 16:50:14 2016 +0100

    ready for 0.1.0

commit cef17cf5a3106a1f65c68e964f982e046777eb10
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Tue Dec 13 15:11:37 2016 +0100

    - readme

commit 15aea698be4b760f845aa49798e25b8209c5d858
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Tue Dec 13 14:56:40 2016 +0100

    - ready for first beta release

commit 8e79ee318013169ca4555b16083d3176b4171e1f
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Mon Dec 5 01:59:00 2016 +0100

    - perl lib dir

commit b8b7fe6b7645ac58d3fed0eb012a1d189cef19e4
Author: Tobias Oetiker <tobi@oetiker.ch>
Date:   Mon Dec 5 01:20:22 2016 +0100

    Auto update (#2)
    
    fix automake

commit e4b3e04e831df8162df4bab9d46f4bd0e56be500
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Mon Dec 5 01:13:07 2016 +0100

    ready for release 0.1.0-rc1

commit 00a04abf5bf0530969d1de5c4beef49bf298e02b
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Mon Dec 5 00:52:33 2016 +0100

    - authors

commit e602980138bacf27eefb84777cacc40434d45200
Author: Tobias Oetiker <tobi@oetiker.ch>
Date:   Mon Dec 5 00:39:29 2016 +0100

    autoconf ahoi

commit 6cb5428eb9f2953aa02810e09b25488412289187
Author: Dominik Hassler <hadfl@cpan.org>
Date:   Sun Dec 4 23:20:37 2016 +0100

    - wip

commit d4b7bb6476a67f07ecd23dc80499426c856610e2
Author: Dominik Hassler <hadfl@users.noreply.github.com>
Date:   Fri Dec 2 14:54:33 2016 +0100

    Initial commit


lib/LXadm/Utils.pm
package LXadm::Utils;

use strict;
use warnings;

use Text::ParseWords qw(shellwords);

my $DLADM    = '/usr/sbin/dladm';
my $IFCONFIG = '/usr/sbin/ifconfig';
my $TEST     = '/usr/bin/test';

# constructor
sub new {
    my $class = shift;
    my $self = { @_ };

    return bless $self, $class
}

# private methods
my $numeric = sub {
    return shift =~ /^\d+$/;
};

my $alphanumeric = sub {
    return shift =~ /^[-\w]+$/;
};

# public methods
sub regexp {
    my $self = shift;
    my $rx = shift;
    my $msg = shift;

    return sub {
        my $value = shift;
        return $value =~ /$rx/ ? undef : "$msg ($value)";
    }
}

sub elemOf {
    my $self = shift;
    my $elems = [ @_ ];

    return sub {
        my $value = shift;
        return (grep { $_ eq $value } @$elems) ? undef
            : 'expected a value from the list: ' . join(', ', @$elems);
    }
}

sub getVNicOver {
    my $self    = shift;
    my $nicName = shift;

    #use string for 'link,over,vid' as perl will warn otherwise
    my @cmd = ($DLADM, qw(show-vnic -p -o), 'link,over');

    open my $vnics, '-|', @cmd or die "ERROR: cannot get vnics\n";

    while (<$vnics>){
        chomp;
        my @nicProps = split ':', $_, 2;
        next if $nicProps[0] ne $nicName;

        return $nicProps[1];
    };

    return undef;
};

sub nicName {
    my $self = shift;

    return sub {
        my ($nicName, $nic) = @_;

        my @cmd;

        # check if VNIC exists
        return undef if $self->getVNicOver($nicName);

        #only reach here if vnic does not exist
        #get first physical link if over is not given
        exists $nic->{over} || do {
            @cmd = ($DLADM, qw(show-phys -p -o link));

            open my $nics, '-|', @cmd or die "ERROR: cannot get nics\n";

            chomp($nic->{over} = <$nics>);
            close $nics;
        };

        @cmd = ($DLADM, qw(create-vnic -l), $nic->{over}, $nicName);
        print STDERR "-> vnic '$nicName' does not exist. creating it...\n";
        system(@cmd) && die "ERROR: cannot create vnic '$nicName'\n";

        return undef;
    }
}

sub purgeVnic {
    my $self = shift;
    my $config = shift;

    for my $nic (@{$config->{net}}){
        my @cmd = ($DLADM, qw(delete-vnic), $nic->{physical});
        system(@cmd) && die "ERROR: cannot delete vnic '$nic->{physical}'\n";
    }
}

1;

__END__

=head1 NAME

LXadm::Utils - lxadm helper module

=head1 SYNOPSIS

use LXadm::Utils;

=head1 DESCRIPTION

methods to check lxadm configuration

=head1 FUNCTIONS

=head2 boolean

checks if the argument is boolean

=head2 numeric

checks if the argument is numeric

=head2 alphanumeric

checks if the argument is alphanumeric

=head2 get_vnic_over

gets the physical nic of a vnic

=head2 nic_name

checks if a vnic exists, tires to create it if not

=head2 serial_name

checks if serial_name is not one of the reserved names

=head2 purge_vnic

deletes all vnic attached to the config

=head1 COPYRIGHT

Copyright 2017 OmniOS Community Edition (OmniOSce) Association.

=head1 LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see L<http://www.gnu.org/licenses/>.

=head1 AUTHOR

S<Dominik Hassler E<lt>hadfl@cpan.orgE<gt>>,

=head1 HISTORY

2016-12-13 had Initial Version

=cut


lib/LXadm/Config.pm
package LXadm::Config;

use strict;
use warnings;

use Illumos::Zones;
use Data::Processor;
use LXadm::Images;
use LXadm::Utils;

# constants/programs
my $TEMPLATE = {
    zonename  => '',
    zonepath  => '',
    brand     => 'lx',
    'ip-type' => 'exclusive',
    net       => [
        {
            physical => '',
            gateway  => '',
            ips      => [
                '',
            ],
            primary  => 'true',
        },
    ],
    'dns-domain' => '',
    resolvers    => [
        '',
    ],
    'kernel-version' => '',
};

my @SPECIALRES = qw(dns-domain resolvers kernel-version);

my $SCHEMA = sub {
    my $sv = LXadm::Utils->new();

    return {
    net     => {
        optional => 1,
        members  => {
            physical => {
                validator => $sv->nicName($sv->nicName(Illumos::Zones->isGZ)),
            },
            over    => {
                optional    => 1,
                description => 'physical nic where vnic traffic goes over',
                example     => '"over" : "igb0"',
                validator   => $sv->regexp(qr/^[-\w]+$/),
            },
        },
    },
    'dns-domain'    => {
        optional    => 1,
        description => 'DNS search domain',
        example     => '"dns-domain" : "example.com"',
        validator   => sub { return undef; },
    },
    resolvers       => {
        optional    => 1,
        array       => 1,
        description => 'DNS resolvers',
        example     => '"resolvers" : [ "8.8.8.8", "8.8.4.4" ]',
        validator   => sub { return undef; },
    },
    'kernel-version' => {
        description => 'Kernel version',
        validator   => sub { return undef; },
    }
    } # return
};

# private methods
my $decodeRes = sub {
    my $self = shift;
    my $cfg  = shift;

    my $schema = $SCHEMA->();

    for (my $i = $#{$cfg->{attr}}; $i >= 0; $i--) {
        my $name = $cfg->{attr}->[$i]->{name};
        next if !grep { $_ eq $name } @SPECIALRES;

        $cfg->{$name} = $schema->{$name}->{array} ? [ split /,/, $cfg->{attr}->[$i]->{value} ]
                      : $cfg->{attr}->[$i]->{value};

        splice @{$cfg->{attr}}, $i, 1;
    }
    # check if attr is empty. if so remove it
    delete $cfg->{attr} if !@{$cfg->{attr}};

    # add VNIC over property
    return if !exists $cfg->{net};
    for my $vnic (@{$cfg->{net}}) {
        my $over = $self->{util}->getVNicOver($vnic->{physical});
        $vnic->{over} = $over if $over;
    }
};

my $encodeRes = sub {
    my $self = shift;
    my $cfg  = shift;

    my $schema = $SCHEMA->();

    $cfg->{attr} //= [];

    for my $res (@SPECIALRES) {
        my %elem = (
            name => $res,
            type => 'string',
        );

        # don't process empty res
        next if !$cfg->{$res};

        $elem{value} = ref $cfg->{$res} eq 'ARRAY' ? join (',', @{$cfg->{$res}})
                     : $cfg->{$res};

        push @{$cfg->{attr}}, { %elem };
        delete $cfg->{$res};
    }

    # remove vnic over property
    delete $_->{over} for (@{$cfg->{net}});
};

# constructor
sub new {
    my $class = shift;
    my $self = { @_ };

    $self->{zone} = Illumos::Zones->new(debug => $self->{debug});
    $self->{imgs} = LXadm::Images->new(debug => $self->{debug});
    $self->{util} = LXadm::Utils->new();

    $self->{cfg}  = Data::Processor->new($self->{zone}->schema); 
    $self->{cfg}->merge_schema($SCHEMA->());

    return bless $self, $class;
}

sub getTemplate {
    my $self = shift;

    return { %$TEMPLATE };
}

sub checkConfig {
    my $self   = shift;
    my $config = shift;

    my $ec = $self->{cfg}->validate($config);
    $ec->count and die join ("\n", map { $_->stringify } @{$ec->{errors}}) . "\n";

    return 1;
}

sub removeLX {
    my $self   = shift;
    my $lxName = shift;
    my $opts   = shift;

    my $config    = $self->readConfig($lxName);
    my $zoneState = $self->{zone}->zoneState($lxName);

    $zoneState eq 'running'
        and die "ERROR: zone '$lxName' still running. use 'lxadm stop $lxName' to stop it first...\n";

    for (keys %$opts) {
        /^vnic$/ && do {
            $self->{util}->purgeVnic($config);
            next;
        };
    }

    $zoneState ne 'configured' && $self->{zone}->uninstallZone($lxName);
    $self->{zone}->deleteZone($lxName);
}

sub writeConfig {
    my $self   = shift;
    my $lxName = shift;
    my $uuid   = shift;
    my $config = shift;

    $self->checkConfig($config);
    $self->$encodeRes($config);

    $self->{zone}->setZoneProperties($lxName, $config, ($uuid ? $self->{imgs}->cachePath . "/$uuid.gz" : ()));

    return 1;
}

sub readConfig {
    my $self   = shift;
    my $lxName = shift;

    my $config = $self->{zone}->getZoneProperties($lxName);
    # check if we are working on a lx zone
    $config->{brand} eq 'lx' or die "ERROR: zone '$lxName' is not a lx zone\n";
    # remove 'rctl' as they are set by aliases
    delete $config->{rctl};
    $self->$decodeRes($config);

    return $config;
}

sub listLX {
    my $self   = shift;
    my $lxName = shift;

    my $lxZones = $lxName ? [ $lxName ] : [ map { $_->{zonename} } @{$self->{zone}->listZones({ brandFilter => qr/^lx$/ })} ];

    # save a copy of $_ in $key as $_ gets modified in function
    return { map { my $key = $_; $key => $self->readConfig($_) } @$lxZones };
}

1;

__END__

=head1 NAME

LXadm::Config - lxadm config class

=head1 SYNOPSIS

use LXadm::Config;

=head1 DESCRIPTION

reads, writes and checkes lxadm configuration

=head1 COPYRIGHT

Copyright 2017 OmniOS Community Edition (OmniOSce) Association.

=head1 LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see L<http://www.gnu.org/licenses/>.

=head1 AUTHOR

S<Dominik Hassler E<lt>hadfl@cpan.orgE<gt>>,

=head1 HISTORY

2016-12-13 had Initial Version

=cut


lib/LXadm/Images.pm
package LXadm::Images;

use strict;
use warnings;

use Data::Dumper;
use JSON;
use File::Path qw(make_path);
use File::Basename qw(dirname);
use File::stat;
use Time::localtime;
use Digest::SHA;

use FindBin;
my ($BASEDIR)   = dirname($FindBin::RealBin);

# constants/programs
my $IMGURL     = 'https://images.joyent.com/images';
my $MAX_AGE    = 24 * 60 * 60;
my $CURL       = '/usr/bin/curl';
my $PKG        = '/usr/bin/pkg';
my $CACHE_PATH = "/var$BASEDIR/cache"; # CACHEPATH
my @DUMPSET    = (
    {
        name  => 'UUID',
        len   => 10,
        key   => [ qw(uuid) ],
        trans => sub { my $str = shift; return substr $str, length ($str) - 8; }
    },
    {
        name  => 'Name',
        len   => 16,
        key   => [ qw(name) ],
    },
    {
        name  => 'Version',
        len   => 10,
        key   => [ qw(version) ],
    },
    {
        name  => 'Kernel',
        len   => 10,
        key   => [ qw(tags kernel_version) ],
    },
    {
        name  => 'Description',
        len   => 50,
        key   => [ qw(description) ],
        trans => sub { return (shift =~ /^[^\s]+\s+(.+\.)\s+Built/)[0]; }
    },
);

# globals
my $images = [];

# private methods
my $getHashValue = sub {
    my $value = shift;
    my $item  = shift;

    $value = $value->{$_} for (@{$item->{key}});

    return $value;
};

my $getImageByUUID = sub {
    my $self = shift;
    my $uuid = shift;

    my $imgs = [ grep { $_->{uuid} =~ /$uuid/ } @{$self->listImages} ];

    @$imgs < 1 and die "ERROR: image UUID containing '$uuid' not found.\n";
    @$imgs > 1 and die "ERROR: more than one image uuid contains '$uuid'.\n";

    return $imgs->[0];
};

my $getFile = sub {
    my $fileName = shift;
    my $url      = shift;

    print "downloading file...\n";
    my @cmd = ($CURL, '-o', "$CACHE_PATH/$fileName", $url);
    system (@cmd) && die "ERROR: cannot download url '$url'.\n";
};

my $checkChecksum = sub {
    my $fileName = shift;
    my $checksum = shift;

    print "checking checksum of '$fileName'...\n";
    if (Digest::SHA->new('sha1')->addfile("$CACHE_PATH/$fileName")->hexdigest
        eq $checksum) {

        return 1;
    }
    print "checksum not ok...\n";
    return 0;
};

my $downloadFile = sub {
    my $fileName = shift;
    my $url      = shift;
    my $opts     = { @_ };

    # check if cache directory exists
    -d "$CACHE_PATH" || make_path("$CACHE_PATH") || do {
        die "ERROR: cannot create directory $CACHE_PATH\n";
    };

    print "checking cache for '$fileName'...\n";
    my $freshDl = 0;
    -f "$CACHE_PATH/$fileName" || do {
        print "file not found in cache...\n";
        $getFile->($fileName, $url);
        $freshDl = 1;
    };

    # check if cache file has a max_age property and redownload if expired
    exists $opts->{max_age} && !$freshDl
        && (time - stat("$CACHE_PATH/$fileName")->mtime > $opts->{max_age})
        && $getFile->($fileName, $url);

    # check checksum if sha1 option is set
    exists $opts->{sha1} && do {
        return if $checkChecksum->($fileName, $opts->{sha1});
        $getFile->($fileName, $url);
        $checkChecksum->($fileName, $opts->{sha1}) || die "ERROR: chechsum mismatch for downloaded file\n";
    };
};

# constructor
sub new {
    my $class = shift;
    my $self = { @_ };

    return bless $self, $class;
}

# public methods
sub cachePath {
    return $CACHE_PATH;
}

sub fetchImages {
    my $self  = shift;
    my $force = shift;

    $downloadFile->('index.json', $IMGURL, (max_age => ($force ? -1 : $MAX_AGE)));
    open my $index, '<', "$CACHE_PATH/index.json" or die "ERROR: cannot open 'index.json': $!\n";
    my $imgs = JSON->new->decode(do { local $/; <$index>; });
    ref $imgs eq 'ARRAY' or die "ERROR: 'index.json' invalid\n";

    # filter so we only get lx-datasets
    $images = [ grep { $_->{type} eq 'lx-dataset' } @$imgs ];
}

sub listImages {
    my $self = shift;
    my $uuid = shift;

    @$images || $self->fetchImages;

    # check if UUID matches a unique item
    defined $uuid && return $self->$getImageByUUID($uuid);

    return $images;
}

sub dumpImages {
    my $self     = shift;
    my $uuid_len = shift;

    my $imgs = $self->listImages;

    print $_->{name} . (' ' x ($_->{len} - length ($_->{name}))) for (@DUMPSET);
    print "\n";
    for my $img (@$imgs) {
        for my $item (@DUMPSET) {
            my $tmpStr = exists $item->{trans} ? $item->{trans}->($getHashValue->($img, $item))
                : $getHashValue->($img, $item);

            $tmpStr = length ($tmpStr) > $item->{len} ? substr ($tmpStr, 0, $item->{len})
                : $tmpStr . (' ' x ($item->{len} - length ($tmpStr)));

            print $tmpStr;
        }
        print "\n";
    }
}

sub downloadImage {
    my $self = shift;
    my $uuid = shift;

    # get image by UUID
    my $img = $self->$getImageByUUID($uuid);

    $downloadFile->("$img->{uuid}.gz", "$IMGURL/$img->{uuid}/file", ( sha1 => $img->{files}->[0]->{sha1} ));

    return $img;
}

1;

__END__

=head1 NAME

LXadm::Images - lxadm images class

=head1 SYNOPSIS

use LXadm::Images;

=head1 DESCRIPTION

downloads and verifies lx zone images

=head1 COPYRIGHT

Copyright 2017 OmniOS Community Edition (OmniOSce) Association.

=head1 LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see L<http://www.gnu.org/licenses/>.

=head1 AUTHOR

S<Dominik Hassler E<lt>hadfl@cpan.orgE<gt>>,

=head1 HISTORY

2016-12-13 had Initial Version

=cut

